rust/agama-lib/src/base_http_client.rs

```diff
@@ -0,0 +1,133 @@
+        let response = self.get_response(path).await?;
+        if response.status().is_success() {
+            response.json::<T>().await.map_err(|e| e.into())
+        } else {
+            Err(self.build_backend_error(response).await)
+        }
+    }
+
+    /// Calls GET method on given path and return Response that can be further
+    /// processed. If only simple object from json is required, use method get.
+    pub async fn get_response(&self, path: &str) -> Result<Response, ServiceError> {
+        self.client
+            .get(self.url(path))
+            .send()
+            .await
+            .map_err(|e| e.into())
+    }
+
+    fn url(&self, path: &str) -> String {
+        self.base_url.clone() + path
+    }
+
+    /// post object to given path and report error if response is not success
+    pub async fn post(&self, path: &str, object: &impl Serialize) -> Result<(), ServiceError> {
+        let response = self.post_response(path, object).await?;
+        if response.status().is_success() {
+            Ok(())
+        } else {
+            Err(self.build_backend_error(response).await)
+        }
+    }
+
+    /// post object to given path and returns server response. Reports error only if failed to send
+    /// request, but if server returns e.g. 500, it will be in Ok result.
+    /// In general unless specific response handling is needed, simple post should be used.
+    pub async fn post_response(
+        &self,
+        path: &str,
+        object: &impl Serialize,
+    ) -> Result<Response, ServiceError> {
+        self.client
+            .post(self.url(path))
+            .json(object)
+            .send()
+            .await
+            .map_err(|e| e.into())
+    }
+
+    /// delete call on given path and report error if failed
+    pub async fn delete(&self, path: &str) -> Result<(), ServiceError> {
+        let response = self.delete_response(path).await?;
+        if response.status().is_success() {
+            Ok(())
+        } else {
+            Err(self.build_backend_error(response).await)
+        }
+    }
+
+    /// delete call on given path and returns server response. Reports error only if failed to send
+    /// request, but if server returns e.g. 500, it will be in Ok result.
+    /// In general unless specific response handling is needed, simple delete should be used.
+    /// TODO: do not need variant with request body? if so, then create additional method.
+    pub async fn delete_response(&self, path: &str) -> Result<Response, ServiceError> {
+        self.client
+            .delete(self.url(path))
+            .send()
+            .await
+            .map_err(|e| e.into())
+    }
+
+    pub async fn build_backend_error(&self, response: Response) -> ServiceError {
+        let code = response.status().as_u16();
+        let text = response
+            .text()
+            .await
+            .unwrap_or_else(|_| Self::NO_TEXT.to_string());
```

Comment:
"No text" -> `"(failed to extract error text from HTTP response)"` is perhaps too verbose but I would appreciate it when debugging it as an operator

Reply:
I check quite deeply when call to text can fail and to be honest I do not find it either in documentation neither from code. So lets use your suggestion.